File: backend/src/services/fraud.service.ts
import Transaction from '../models/Transaction';
import FraudAlert from '../models/FraudAlert';
import { logger } from '../utils/logger';
import mongoose from 'mongoose';

export const detectFraud = async (userId: string, transactionId: string) => {
  try {
    const tx = await Transaction.findById(transactionId);
    if (!tx) return;

    const uid = new mongoose.Types.ObjectId(userId);
    const fiveMinutesAgo = new Date(Date.now() - 5 * 60 * 1000);

    // Rule 1: Rapid transactions (> 10 in 5 minutes)
    const recentCount = await Transaction.countDocuments({
      userId: uid,
      createdAt: { $gte: fiveMinutesAgo },
    });
    if (recentCount > 10) {
      await createAlert(userId, transactionId, 'rapid_transactions', 'high', `${recentCount} transactions in 5 minutes`);
    }

    // Rule 2: Unusual amount (5x average)
    const avgResult = await Transaction.aggregate([
      { $match: { userId: uid, type: tx.type } },
      { $group: { _id: null, avg: { $avg: '$amount' } } },
    ]);
    const avg = avgResult[0]?.avg || 0;
    if (avg > 0 && tx.amount > avg * 5) {
      await createAlert(userId, transactionId, 'unusual_amount', 'medium', `Amount ₦${tx.amount} is 5x the average ₦${Math.round(avg)}`);
    }
  } catch (err) {
    logger.error('Fraud detection error:', err);
  }
};

const createAlert = async (userId: string, transactionId: string, alertType: string, severity: string, description: string) => {
  await FraudAlert.create({ userId, transactionId, alertType, severity, description });
  await Transaction.findByIdAndUpdate(transactionId, { fraudFlag: true, fraudReason: description });
};

File: ai-service/app/services/fraud_service.py
from typing import List, Optional
import structlog

logger = structlog.get_logger()


def detect_fraud(transaction: dict, recent_transactions: List[dict]) -> Optional[dict]:
    """Run rule-based fraud detection on a transaction."""
    alerts = []

    # Rule 1: Rapid transactions > 10 in 5 minutes
    if len(recent_transactions) > 10:
        alerts.append({
            "alert_type": "rapid_transactions",
            "severity": "high",
            "description": f"{len(recent_transactions)} transactions recorded in 5 minutes",
        })

    # Rule 2: Amount is 5x the user's average
    if recent_transactions:
        avg = sum(t["amount"] for t in recent_transactions) / len(recent_transactions)
        if avg > 0 and transaction["amount"] > avg * 5:
            alerts.append({
                "alert_type": "unusual_amount",
                "severity": "medium",
                "description": f"Amount ₦{transaction['amount']:,} is {transaction['amount']/avg:.1f}x the average ₦{avg:,.0f}",
            })

    # Rule 3: Negative or zero inventory after sale
    if transaction.get("type") == "sale" and transaction.get("resulting_stock") is not None:
        if transaction["resulting_stock"] < 0:
            alerts.append({
                "alert_type": "inventory_mismatch",
                "severity": "medium",
                "description": "Sale recorded would result in negative inventory",
            })

    return alerts[0] if alerts else None
